### PR TITLE
Expose dimRoundingMode attribute of pool op

### PR DESCRIPTION
**Pull request by BruceDai**
Link to https://github.com/webmachinelearning/webnn-polyfill/issues/135.

@lina128 PTAL, thanks.

---
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/5849)

### CODE REVIEW

**Review by lina128 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/5849)* status: 0 of 1 approvals obtained (waiting on @BruceDai)

___
*[tfjs-core/src/ops/avg_pool.ts, line 51 at r1](https://reviewable.io/reviews/tensorflow/tfjs/5849#-MoZlZWF5G5gbeWdF-i5:-MoZlZWF5G5gbeWdF-i6:b-ez8jl1) ([raw file](https://github.com/tensorflow/tfjs/blob/c559d768326c16d6d5f8a3c6c132a81f62728efb/tfjs-core/src/ops/avg_pool.ts#L51)):*
> ```
>  * @param outputSizes The height and width of output shape:
>  *     `[outHeight, outWidth]`.
> ```
The use case is not clear from this description. Can you elaborate the use case here?

**Comment by BruceDai** [comment]
> The use case is not clear from this description. Can you elaborate the use case here?

Hi @lina128, as you known, [WebNN-polyfill](https://github.com/webmachinelearning/webnn-polyfill) is a JavaScript implementation of the [Web Neural Network API](https://webmachinelearning.github.io/webnn/), WebNN-polyfill now uses tf.js to implement op functions. Recently pool2d operations of  WebNN API enabled new `roundingType` and `outputSizes` options, the `roundingType` could map to `dimRoundingMode`,  and here's WebNN API - pool2d's  `outputSizes` description

```
outputSizes: a sequence of long of length 2. The sizes of the two spacial dimensions of the output tensor. 
When the output sizes are explicitly specified, the options.roundingType is ignored. 
If not specified, the output sizes are automatically computed.
``` 

In tf.js, the output shape of pool operations currently are automatically computed. This proposal of adding `outputSizes` attribute to specify output-height and output-width is for the use case of WebNN API - pool2d supporting  `outputSizes`  option. Thanks.

**Review by pyu10055 — CHANGES_REQUESTED** [review]
Thanks 

> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/5849#-:-Mob7gA56jQjQpgT6o9l:b-ucfkni)* status: 0 of 1 approvals obtained (waiting on @BruceDai and @lina128)

___
*[tfjs-core/src/ops/avg_pool.ts, line 51 at r1](https://reviewable.io/reviews/tensorflow/tfjs/5849#-MoZlZWF5G5gbeWdF-i5:-Mob8TchC5x54MWhwVNP:b9q09ql) ([raw file](https://github.com/tensorflow/tfjs/blob/c559d768326c16d6d5f8a3c6c132a81f62728efb/tfjs-core/src/ops/avg_pool.ts#L51)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

> ```
>  * @param outputSizes The height and width of output shape:
>  *     `[outHeight, outWidth]`.
> ```
The use case is not clear from this description. Can you elaborate the use case here?
</blockquote></details>

might be good to copy the description of these two field from the webnn page
https://github.com/webmachinelearning/webnn/pull/208/files#diff-5e793325cd2bfc452e268a4aa2f02b4024dd9584bd1db3c2595f61f1ecf7b985R1604
___
*[tfjs-core/src/ops/avg_pool.ts, line 82 at r1](https://reviewable.io/reviews/tensorflow/tfjs/5849#-Mob6h-U7T1S3XRwT1Yy:-Mob6h-U7T1S3XRwT1Yz:bbtjjmw) ([raw file](https://github.com/tensorflow/tfjs/blob/c559d768326c16d6d5f8a3c6c132a81f62728efb/tfjs-core/src/ops/avg_pool.ts#L82)):*
> ```XML
>     if (typeof pad === 'string'){
>       throw new Error(
>           `dimRoundingMode ${dimRoundingMode} is set when using ${pad} pad`);
> ```

the error message is showing the has happened, but not explain why the dimRoundingMode can not be set with the pad value.
___
*[tfjs-core/src/ops/conv_util.ts, line 366 at r1](https://reviewable.io/reviews/tensorflow/tfjs/5849#-Mob9Vpy2-d_2h_7MDgY:-Mob9Vpy2-d_2h_7MDgZ:b61gzq5) ([raw file](https://github.com/tensorflow/tfjs/blob/c559d768326c16d6d5f8a3c6c132a81f62728efb/tfjs-core/src/ops/conv_util.ts#L366)):*
> ```XML
>     for (const t of [undefined, 'floor', 'round', 'ceil']) {
>       [outputRows, outputCols] =
>           computeOutputShape2D(inShape, fieldSize, stride, zeroPad,
> ```

the check should be done if `roundingMode === null`, otherwise should verify if the otuputSize matches with the provided roundingModel value.

**Comment by BruceDai** [comment]
Thanks @pyu10055 

> the check should be done if roundingMode === null, otherwise should verify if the otuputSize matches with the provided roundingModel value.

I added the explicit log information of **The dimRoundingMode is ignored when the output sizes are explicitly specified.** Is it ok for not checking given roundingMode with given otuputSize?

@pyu10055 @lina128 Please take another look at the updated commit, thanks.

**Comment by huningxin** [comment]
Thanks for review @pyu10055 and @lina128 .

The output sizes of WebNN is to be compatible to native APIs that don't support setting dim rounding mode. As TF.js already supports dim rounding mode, I don't think it needs to support the output sizes. The WebNN-poyfill can translate the output sizes to TF.js dim rounding mode. 

@BruceDai , if it makes sense, please focus on exposing dimRoundingMode of TF.js pool op. Thanks.

**Comment by BruceDai** [comment]
Thanks @huningxin.

Yes, it makes sense. I will update this pr,  then ping @pyu10055 and @lina128 for review.

**Comment by BruceDai** [comment]
@pyu10055 @lina128 I updated commit according to @huningxin's comment, PTAL, thanks.

**Review by pyu10055 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/5849)* status: 0 of 1 approvals obtained (waiting on @BruceDai and @lina128)

___
*[tfjs-core/src/ops/max_pool.ts, line 81 at r3](https://reviewable.io/reviews/tensorflow/tfjs/5849#-MoiuPG0-lg2gUtDPuCg:-MoiuPG1-utDapE7h3Nz:bghfhb4) ([raw file](https://github.com/tensorflow/tfjs/blob/424c41c8edc735f4adc418e745bc8338575646c9/tfjs-core/src/ops/max_pool.ts#L81)):*
> ```XML
>     if (typeof pad === 'string') {
>       throw Error(
>           `Error in maxPool: pad must be an integer when using `  +
> ```

this error message has been repeated through multiple ops, it would be good to refactor it into a help method and share across all ops.
___
*[tfjs-core/src/ops/max_pool_test.ts, line 114 at r3](https://reviewable.io/reviews/tensorflow/tfjs/5849#-Moiuatm3IoT2tbZMn2f:-Moiuatm3IoT2tbZMn2g:b-6d5yzw) ([raw file](https://github.com/tensorflow/tfjs/blob/424c41c8edc735f4adc418e745bc8338575646c9/tfjs-core/src/ops/max_pool_test.ts#L114)):*
> ```XML
>   });
> 
>   it('x=[2,2,3] f=[2,2] s=3 p=1 default dimRoundingMode', () => {
> ```

please add failure test across all ops.

**Comment by BruceDai** [comment]
> this error message has been repeated through multiple ops, it would be good to refactor it into a help method and share across all ops.
> please add failure test across all ops.

Thanks @pyu10055

I've refactored pad checking when using `dimRoundingMode`. I'm ongoing to write test cases for those ops, new commit will be submitted later, thanks.

**Comment by BruceDai** [comment]
@pyu10055 @lina128 Please take a look again, thanks.

**Review by pyu10055 — APPROVED** [review]
Great work, thanks!

> Reviewed 15 of 20 files at r3, 25 of 25 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/5849#-:-MpI9lst7Dp0EEEs6RDb:brhzsor)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @BruceDai and @lina128)

**Comment by BruceDai** [comment]
@lina128 I rebased code, PTAL, thanks.

**Review by lina128 — APPROVED** [review]
LGTM, thanks!

> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/5849#-:-Mpln6aeEl98FpZLLiLE:b2c0zcb)* status: :shipit: complete! 2 of 1 approvals obtained (waiting on @BruceDai and @lina128)
